File: conformance/go/pcm.go
package conformance

// The sample anchor, FORMAT.md 6.2, parsed by hand from the RIFF bytes.

import (
	"encoding/binary"
	"fmt"
)

// PcmHash returns the anchor and true, or "" and false when the bytes are
// not a WAV this specification defines an anchor for.
func PcmHash(b []byte) (string, bool) {
	if len(b) < 12 || string(b[0:4]) != "RIFF" || string(b[8:12]) != "WAVE" {
		return "", false
	}
	var channels, bits uint16
	var rate uint32
	var haveFmt bool
	var data []byte
	at := 12
	for at+8 <= len(b) {
		id := string(b[at : at+4])
		size := int(binary.LittleEndian.Uint32(b[at+4 : at+8]))
		body := at + 8
		if body+size > len(b) {
			return "", false
		}
		if id == "fmt " {
			if size < 16 {
				return "", false
			}
			tag := binary.LittleEndian.Uint16(b[body : body+2])
			if tag != 1 && tag != 3 {
				return "", false
			}
			channels = binary.LittleEndian.Uint16(b[body+2 : body+4])
			rate = binary.LittleEndian.Uint32(b[body+4 : body+8])
			bits = binary.LittleEndian.Uint16(b[body+14 : body+16])
			haveFmt = true
		}
		if id == "data" {
			data = b[body : body+size]
		}
		at = body + size + size%2
	}
	if !haveFmt || data == nil {
		return "", false
	}
	prefix := []byte(fmt.Sprintf("pcm|%d|%d|%d|", channels, rate, bits))
	return sha256Hex(append(prefix, data...)), true
}

// FindAnchors and FindPcmAnchors, FORMAT.md 6.3.
func FindAnchors(sha string, entries []map[string]any) []map[string]any {
	return matching(entries, "sha256", sha)
}

func FindPcmAnchors(pcm string, entries []map[string]any) []map[string]any {
	if pcm == "" {
		return nil
	}
	return matching(entries, "pcm_sha256", pcm)
}

func matching(entries []map[string]any, key, want string) []map[string]any {
	var out []map[string]any
	for _, e := range entries {
		if a, ok := e["artifact"].(map[string]any); ok && a != nil {
			if s, ok := a[key].(string); ok && s == want {
				out = append(out, e)
			}
		}
	}
	return out
}

File: src/pcm.js
/* The PCM secondary anchor: a hash of the audio's sample data plus the
   parameters needed to play it — not the file's bytes. Retagging a WAV
   (cover art, INFO chunks, bext) changes the file hash; it cannot change
   this one. Deterministic by construction: we parse the container
   ourselves, byte for byte — no Web Audio, no decoder variance — so the
   same recording hashes the same in every browser. Uncompressed WAV
   (PCM and IEEE float) only in phase 1; anything else is null, honestly. */

const HEX = (buf) => [...new Uint8Array(buf)].map((b) => b.toString(16).padStart(2, "0")).join("");

const fourcc = (view, at) =>
  String.fromCharCode(view.getUint8(at), view.getUint8(at + 1), view.getUint8(at + 2), view.getUint8(at + 3));

/* Returns the hex SHA-256 of "pcm|channels|rate|bits|" + data bytes, or
   null when the buffer is not a WAV this phase understands. */
export const pcmHash = async (buffer) => {
  try {
    const view = new DataView(buffer);
    if (view.byteLength < 12 || fourcc(view, 0) !== "RIFF" || fourcc(view, 8) !== "WAVE") return null;

    let fmt = null;
    let data = null;
    let at = 12;
    while (at + 8 <= view.byteLength) {
      const id = fourcc(view, at);
      const size = view.getUint32(at + 4, true);
      const body = at + 8;
      if (body + size > view.byteLength) return null; /* truncated chunk */
      if (id === "fmt ") {
        if (size < 16) return null;
        const format = view.getUint16(body, true);
        if (format !== 1 && format !== 3) return null; /* PCM & IEEE float only, for now */
        fmt = {
          channels: view.getUint16(body + 2, true),
          rate: view.getUint32(body + 4, true),
          bits: view.getUint16(body + 14, true),
        };
      }
      if (id === "data") data = new Uint8Array(buffer, body, size);
      at = body + size + (size % 2); /* RIFF pads odd chunks */
    }
    if (!fmt || !data) return null;

    const prefix = new TextEncoder().encode(`pcm|${fmt.channels}|${fmt.rate}|${fmt.bits}|`);
    const material = new Uint8Array(prefix.length + data.length);
    material.set(prefix, 0);
    material.set(data, prefix.length);
    return HEX(await crypto.subtle.digest("SHA-256", material));
  } catch {
    return null;
  }
};

export const pcmHashFile = async (file) => pcmHash(await file.arrayBuffer());
